商品出品機能

nxu4559 opened the pull request:
#What
商品出品機能

#Why
フリマアプリの商品が出品できる機能

nxu4559 committed:
itemモデルとテーブル作成

nxu4559 committed:
Active Storage導入

nxu4559 committed:
Active Hash導入

nxu4559 committed:
商品出品ページ表示

nxu4559 committed:
価格表示

nxu4559 committed:
テストコード

商品出品の単体テストコード

nxu4559 committed:
rubocop item_spec.rb

nxu4559 commented:
以下が実際の動作の様子です。
https://gyazo.com/829101c893cca1dd55e8b9f37af1c6ca

mizuki-iio commented on app/controllers/items_controller.rb:
```diff
@@ -1,10 +1,29 @@
+  before_action :authenticate_user!, only: [:new]
 
   def index
+    @items = Item.order('created_at DESC')
```
こちらのindexアクションの記述は、商品一覧機能実装時に記述していただくため、現段階では削除またはコメントアウトしておきましょう！

mizuki-iio commented on app/views/items/edit.html.erb:
```diff
@@ -7,10 +7,10 @@ app/assets/stylesheets/items/new.css %>
   <div class="items-sell-main">
     <h2 class="items-sell-title">商品の情報を入力</h2>
-    <%= form_with local: true do |f| %>
+    <%= form_with model: @item local: true do |f| %>
```
こちらのedit.html.erbの記述は商品編集機能実装時に記述していただくため、現段階では修正前の状態に戻しておきましょう！

mizuki-iio commented on app/views/items/index.html.erb:
```diff
@@ -125,13 +125,14 @@
     <%= link_to '新規投稿商品', "#", class: "subtitle" %>
     <ul class='item-lists'>
-
+      
```
こちらのindex.html.erbの記述は、商品一覧機能実装時に修正していただくため、現段階では修正前の状態に戻しておきましょう！

mizuki-iio commented on app/views/items/show.html.erb:
```diff
@@ -4,10 +4,10 @@
   <div class="item-box">
     <h2 class="name">
-      <%= "商品名" %>
+      <%= @item.name %>
```
こちらのshow.html.erbの記述は、商品詳細機能実装時に修正していただくため、現段階では修正前の状態に戻しておきましょう！

mizuki-iio commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,90 @@
+          end
+
+          it 'priceが300円以上でないと保存できない' do
+            @item.price = '299'
```
こちらと一つ下の、価格についての2テストは、数値でのテストを行うため、「299」のようにシングルクオーテーションを外して記述しましょう！

mizuki-iio commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,90 @@
+            expect(@item.errors.full_messages).to include("Description can't be blank")
+          end
+
+          it 'categoryは１を登録したら保存できない' do
```
active hashのカラムのテストについては、「1のとき」の他に「nilのとき」も実装しましょう！
active hashのカラム全てが対象になります！

mizuki-iio reviewed (commented):
レビュー依頼ありがとうございます！
6点コメントいたしましたので、ご確認をお願いいたします！

初めの方で指摘させていただきましたが、前提として「一つのプルリクエストでは一つの機能の実装」をしていただく必要があります。
他の機能の実装まで、このプルリクエストで行ってしまうと、その機能のプルリクエストの段階で差分としてあがらず、レビューが行えなくなってしまうためです。
次回以降こちらをご認識いただいた上で、取り組んでいただければと思います！

また、以下の挙動が確認できるgyazoを添付していただけますでしょうか？
```
出品に失敗しエラーが表示される動画
ログインしていないユーザーが出品しようとするとログインページへリダイレクトされる動画
テスト結果の画像
```

nxu4559 committed:
商品出品機能　修正1

nxu4559 commented:
動作の様子
・出品に失敗しエラーが表示される動画
・ログインしていないユーザーが出品しようとするとログインページへリダイレクトされる動画
・テスト結果の画像
https://gyazo.com/8087e52c9381c82641207b2bde78228e
https://gyazo.com/f8d6592f7921b4fb82c8b66b651e94d7
https://gyazo.com/05c4331245ba0472f108fa6ac4f7398c

chirikuzu commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,120 @@
+          end
+
+          it 'categoryは１を登録したら保存できない' do
+            @item.category_id = '1'
```
```ruby
@item.category_id = '1'
```
こちらですが、文字列として認識されてしまいますので、シングルクォーテーションを外してあげましょう。
商品の状態、発送までの日数、配送料の負担、発送元地域などについても同様です。

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします。

nxu4559 committed:
商品出品機能　修正２

テストコードの修正
''をつけると文字列として認識されるためこれらを外した

masaya-sakurai commented on spec/factories/categories.rb:
```diff
@@ -0,0 +1,4 @@
+FactoryBot.define do
+  factory :category do
+  end
+end
```
こちら今回はテストに必要のないファイルとなります！不必要なファイルは可読性向上のために削除いたしましょう！
今回ですと、以下のものが該当するかと思います！それぞれにコメントを残しておきましたので確認お願いいたします。

 spec/factories/categories.rb
 spec/factories/conditions.rb 
 spec/factories/days.rb
 spec/factories/postage_prayers.rb
 spec/factories/prefectures.rb
 spec/models/category_spec.rb
 spec/models/condition_spec.rb
 spec/models/day_spec.rb
 spec/models/postage_prayer_spec.rb
 spec/models/prefecture_spec.rb

masaya-sakurai commented on spec/factories/conditions.rb:
```diff
@@ -0,0 +1,4 @@
+FactoryBot.define do
+  factory :condition do
+  end
+end
```
こちらも削除しましょう！

masaya-sakurai commented on spec/factories/days.rb:
```diff
@@ -0,0 +1,4 @@
+FactoryBot.define do
+  factory :day do
+  end
+end
```
こちらも削除しましょう！

masaya-sakurai commented on spec/factories/postage_prayers.rb:
```diff
@@ -0,0 +1,4 @@
+FactoryBot.define do
+  factory :postage_prayer do
+  end
+end
```
こちらも削除しましょう！

masaya-sakurai commented on spec/factories/prefectures.rb:
```diff
@@ -0,0 +1,4 @@
+FactoryBot.define do
+  factory :prefecture do
+  end
+end
```
こちらも削除しましょう！

masaya-sakurai commented on spec/models/category_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Category, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
こちらも削除しましょう！

masaya-sakurai commented on spec/models/condition_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Condition, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
こちらも削除しましょう！

masaya-sakurai commented on spec/models/day_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Day, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
こちらも削除しましょう！

masaya-sakurai commented on spec/models/postage_prayer_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe PostagePrayer, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
こちらも削除しましょう！

masaya-sakurai commented on spec/models/prefecture_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Prefecture, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
こちらも削除しましょう！

masaya-sakurai reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認いたしました！

追加でいくつかコメントさせていただきましたので、ご確認お願いいたします！
あと少しですね！頑張りましょう！

nxu4559 committed:
商品出品機能　修正3

spec/factories/categories.rb
spec/factories/conditions.rb
spec/factories/days.rb
spec/factories/postage_prayers.rb
spec/factories/prefectures.rb
spec/models/category_spec.rb
spec/models/condition_spec.rb
spec/models/day_spec.rb
spec/models/postage_prayer_spec.rb
spec/models/prefecture_spec.rb削除

masaya-sakurai commented:
レビュー依頼ありがとうございます！
修正点確認いたしました！
LGTMとさせていただきます！実装お疲れさまでした！
最後にMergeとPullを忘れないようにお願いいたします！

次は商品一覧機能ですね！頑張ってください！
